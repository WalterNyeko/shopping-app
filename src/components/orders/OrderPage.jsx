import React, { Fragment } from "react";
import { NavLink } from "react-router-dom";

const OrderPage = ({
  cartItems,
  totalAmount,
  handleClick,
  handleClickRemoveItem,
  handleClickIncreaseItemQuantity,
  handleClickDecreaseItemQuantity
}) => {
  return (
    <Fragment>
      <section className="container content">
        <div className="row mt-5 mb-5">
          <div className="col-md-2 col-sm-12">
            <button className="btn btn-danger btn-block" onClick={handleClick}>
              EMPTY CART
            </button>
          </div>
          <div className="col-md-8 col-sm-12 text-center">
            <h4 className="text-danger pt-3">
              <strong>Total: $ {totalAmount ? totalAmount : "0.0"}</strong>
            </h4>
          </div>
          <div className="col-md-2 col-sm-12">
            {cartItems.length ? (
              <NavLink to="/orders" className="btn btn-danger">
                PLACE ORDER
              </NavLink>
            ) : (
              <NavLink to="/orders" className="btn btn-danger">
                MY ORDERS
              </NavLink>
            )}
          </div>
        </div>
        <div className="table-responsive">
          <table className="table table-striped table-hover">
            <thead>
              <tr>
                <th />
                <th>Name</th>
                <th>Attribute</th>
                <th>Price</th>
                <th>Quantity</th>
                <th>Subtotal</th>
              </tr>
            </thead>
            <tbody>
              {cartItems.length
                ? cartItems.map(
                    ({
                      item_id,
                      name,
                      price,
                      subtotal,
                      attributes,
                      quantity
                    }) => (
                      <tr key={item_id}>
                        <td>
                          <span
                            className="remove-order"
                            onClick={() => handleClickRemoveItem(item_id)}
                          >
                            <i className="fas fa-times text-danger" /> Remove
                          </span>
                        </td>
                        <td>{name}</td>
                        <td>{attributes}</td>
                        <td>{price}</td>
                        <td>
                          <button
                            className="btn btn-danger btn-sm btn-group"
                            onClick={() =>
                              handleClickDecreaseItemQuantity(item_id)
                            }
                          >
                            -
                          </button>
                          <button className="btn btn-sm btn-group">
                            {quantity}
                          </button>
                          <button
                            className="btn btn-danger btn-sm btn-group"
                            onClick={() =>
                              handleClickIncreaseItemQuantity(item_id)
                            }
                          >
                            +
                          </button>
                        </td>
                        <td>${subtotal}</td>
                      </tr>
                    )
                  )
                : ""}
            </tbody>
          </table>
        </div>
      </section>
    </Fragment>
  );
};
export default OrderPage;
