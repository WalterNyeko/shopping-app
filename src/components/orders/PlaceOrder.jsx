import React from "react";
import StripeCheckout from "react-stripe-checkout";
import { FETCH_CART_ITEMS } from "../../store/types";

const PlaceOrder = ({
  state,
  shippingRegions,
  handleInputChange,
  handlePlaceOrder,
  myOrders,
  handleFetchOrderById,
  orderOfParticularId,
  orderDetails,
  formatDate,
  handlePayment,
  email,
  cartItems
}) => {
  return (
    <section className="container content mb-5">
      <div className="row">
        <div className="col-md-6">
          {cartItems.length ? (
            <div className="card">
              <div className="card-header">Place Order</div>
              <div className="card-body">
                <form>
                  <div className="form-group row">
                    <label
                      for="shipping-address"
                      className="col-md-4 text-right"
                    >
                      Ship To:
                    </label>
                    <select
                      className="form-control col-md-7"
                      onChange={handleInputChange}
                      name="shipping_id"
                    >
                      {shippingRegions &&
                        shippingRegions.length &&
                        shippingRegions.map(
                          ({ index, shipping_region_id, shipping_region }) => (
                            <option value={shipping_region_id} key={index}>
                              {shipping_region}
                            </option>
                          )
                        )}
                    </select>
                    {state.errors.shipping_id && (
                      <span className="text-danger">
                        {state.errors.shipping_id}
                      </span>
                    )}
                  </div>

                  <div className="form-group row">
                    <label for="tax" className="col-md-4 text-right">
                      Tax Type:
                    </label>
                    <label for="tax" className="col-md-7 text-right">
                      {state.tax_type}
                    </label>
                  </div>
                  <div className="form-group row">
                    <label for="tax_type" className="col-md-4 text-right">
                      Tax Percentage:
                    </label>
                    <label for="tax_type" className="col-md-7 text-right">
                      {state.tax_percentage}
                    </label>
                  </div>
                  <div className="form-group row">
                    <div className="col-md-4" />
                    <div className="col-md-7 p-0">
                      <button
                        type="submit"
                        className="btn btn-danger btn-block"
                        onClick={handlePlaceOrder}
                      >
                        Place Order
                      </button>
                    </div>
                  </div>
                </form>
              </div>
            </div>
          ) : (
            ""
          )}

          <div className={`order-table ${cartItems.length ? "mt-5" : ""}`}>
            <div className="card">
              <div className="card-header">My Orders</div>
              <div className="card-body">
                <table className="table table-striped table-hover table-sm">
                  <thead>
                    <tr>
                      <th>Ordered By</th>
                      <th>Shipped On</th>
                      <th>Status</th>
                      <th>Ordered On</th>
                      <th>Total Amount</th>
                    </tr>
                  </thead>
                  <tbody>
                    {myOrders &&
                      myOrders.length &&
                      myOrders.map(
                        ({
                          name,
                          order_id,
                          shipped_on,
                          status,
                          total_amount,
                          created_on
                        }) => (
                          <tr
                            onClick={() => handleFetchOrderById(order_id)}
                            key={order_id}
                          >
                            <td>{name}</td>
                            <td>{shipped_on ? shipped_on : "Not shipped"}</td>
                            <td>{status === 0 ? "Not paid" : "Paid"}</td>
                            <td>{formatDate(created_on)}</td>
                            <td>{`$ ${total_amount}`}</td>
                          </tr>
                        )
                      )}
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
        <div className="col-md-6 mx-auto">
          <div className="card">
            <div className="card-header">Order Summary</div>
            <div className="card-body">
              <table className="table table-striped table-hover table-sm">
                <thead>
                  <tr>
                    <th>Product</th>
                    <th>Attributes</th>
                    <th>Quantity</th>
                    <th>Unit Cost</th>
                    <th>SubTotal</th>
                  </tr>
                </thead>
                <tbody>
                  {orderOfParticularId &&
                    orderOfParticularId.length &&
                    orderOfParticularId.map(
                      ({
                        product_name,
                        attributes,
                        quantity,
                        unit_cost,
                        subtotal
                      }) => (
                        <tr>
                          <td>{product_name}</td>
                          <td>{attributes}</td>
                          <td>{quantity}</td>
                          <td>{`$ ${unit_cost}`}</td>
                          <td>{`$ ${subtotal}`}</td>
                        </tr>
                      )
                    )}
                </tbody>
              </table>

              <form>
                <div className="form-group">
                  <hr className="m-0" />
                  <div className="order-total pt-2">
                    <h5>
                      Order Total Cost{" "}
                      <span className="float-right">{`$ ${
                        orderDetails.total_amount
                          ? orderDetails.total_amount
                          : "0"
                      }`}</span>
                    </h5>
                  </div>
                </div>
              </form>
            </div>
            <div className="card-footer">
              <StripeCheckout
                token={handlePayment}
                stripeKey="pk_test_a7BkMHivjiRVLnCVyXh3SkzO00UdQn0SYm"
                amount={orderDetails.total_amount * 100} // cents
                label="PAY NOW"
                billingAddress
                shippingAddress
                email={email}
              />
            </div>
          </div>
        </div>
      </div>
    </section>
  );
};

export default PlaceOrder;
