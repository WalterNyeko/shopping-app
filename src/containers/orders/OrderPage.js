import React, { Component, Fragment } from "react";
import OrderPageComponent from "../../components/orders/OrderPage";
import UpperNav from "../commons/UpperNav";
import Footer from "../commons/Footer";
import { connect } from "react-redux";
import { getDepartments } from "../../store/actions/Departments";
import {
  getItemsInCart,
  getTotalAmount,
  emptyShoppingCart,
  deleteItemFromShoppingCart,
  createChargeOnCard,
  editItemFromShoppingCart
} from "../../store/actions/Orders";

export class OrderPage extends Component {
  componentWillMount = () => {
    const { getItemsInCart, getDepartments, getTotalAmount } = this.props;
    getDepartments();
    const cartId = localStorage.getItem("cartId");
    cartId && getItemsInCart(cartId);
    cartId && getTotalAmount(cartId);
  };

  /**
   * ensures that a user can delete an item from the cart
   *
   * @returns {void}
   */
  handleClickRemoveItem = itemId => {
    const cartId = localStorage.getItem("cartId");
    const { deleteItemFromShoppingCart } = this.props;
    deleteItemFromShoppingCart(cartId, itemId);
  };

  /**
   * ensures that a user can increase quantity an item from the cart
   *
   * @returns {void}
   */
  handleClickIncreaseItemQuantity = itemId => {
    const cartId = localStorage.getItem("cartId");
    const { editItemFromShoppingCart } = this.props;
    const theSelectedObject = this.props.myCart.cartItems.filter(
      item => item.item_id === itemId
    );
    let { quantity } = theSelectedObject[0];
    quantity++;
    const data = {
      item_id: itemId,
      quantity
    };
    editItemFromShoppingCart(cartId, data);
  };

  /**
   * ensures that a user can decrease quantity an item from the cart
   *
   * @returns {void}
   */
  handleClickDecreaseItemQuantity = itemId => {
    const cartId = localStorage.getItem("cartId");
    const { editItemFromShoppingCart } = this.props;
    const theSelectedObject = this.props.myCart.cartItems.filter(
      item => item.item_id === itemId
    );
    let { quantity } = theSelectedObject[0];
    quantity--;
    const data = {
      item_id: itemId,
      quantity
    };
    editItemFromShoppingCart(cartId, data);
  };

  /**
   * ensures the user's cart is emptied
   *
   * @returns {void}
   */
  handleEmptyShoppingCart = () => {
    const { emptyShoppingCart } = this.props;
    const cartId = localStorage.getItem("cartId");
    emptyShoppingCart(cartId);
  };

  render() {
    const { departments } = this.props.allDepartments;
    const {
      cartItems,
      totalAmountOfItemsInCart: { total_amount }
    } = this.props.myCart;

    return (
      <Fragment>
        <UpperNav />
        <OrderPageComponent
          cartItems={cartItems}
          totalAmount={total_amount}
          handleClick={this.handleEmptyShoppingCart}
          handlePayment={this.handlePayment}
          handleClickRemoveItem={this.handleClickRemoveItem}
          handleClickIncreaseItemQuantity={this.handleClickIncreaseItemQuantity}
          handleClickDecreaseItemQuantity={this.handleClickDecreaseItemQuantity}
        />
        <Footer departments={departments} />
      </Fragment>
    );
  }
}
const mapStateToProps = state => ({
  allDepartments: state.departments,
  myCart: state.itemsInCart
});
export default connect(
  mapStateToProps,
  {
    getDepartments,
    getItemsInCart,
    getTotalAmount,
    emptyShoppingCart,
    deleteItemFromShoppingCart,
    createChargeOnCard,
    editItemFromShoppingCart
  }
)(OrderPage);
