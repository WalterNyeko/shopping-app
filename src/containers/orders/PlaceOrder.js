import React, { Fragment } from "react";
import { connect } from "react-redux";
import PlaceOrderComponent from "../../components/orders/PlaceOrder";
import UpperNav from "../../components/commons/UpperNav";
import Footer from "../../components/commons/Footer";
import { getDepartments } from "../../store/actions/Departments";
import {
  getShippingRegions,
  getTaxes,
  getCustomersDetails
} from "../../store/actions/Customers";
import {
  placeOrder,
  getAllMyOrders,
  getOrderById,
  getOrderDetailById,
  getTotalAmount,
  getItemsInCart,
  createChargeOnCard
} from "../../store/actions/Orders";
import history from "../../helpers/history";
import { showErrorNotification } from "../../helpers/index";

export class PlaceOrder extends React.Component {
  state = {
    isLoggedIn: false,
    shipping_id: 0,
    errors: {},
    tax_id: 0,
    tax_type: "",
    tax_percentage: 0,
    orderId: 0
  };
  componentWillMount() {
    const {
      getDepartments,
      getShippingRegions,
      getTaxes,
      getAllMyOrders,
      getTotalAmount,
      getCustomersDetails
    } = this.props;
    const cartId = localStorage.getItem("cartId");
    getDepartments();
    const jwtToken = localStorage.getItem("jwt-token");
    jwtToken && this.setState({ isLoggedIn: true });
    getShippingRegions();
    getTaxes();
    getAllMyOrders();
    cartId && getTotalAmount(cartId);
    cartId && getItemsInCart(cartId);
    getCustomersDetails();
  }

  handleInputChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value }, () => this.populateStateImmediately());
  };

  populateStateImmediately = () => {
    const { shipping_id } = this.state;
    if (shipping_id === "1" || shipping_id === "2") {
      const {
        tax_id,
        tax_type,
        tax_percentage
      } = this.props.theRegions.taxes[1];
      this.setState({ tax_id, tax_type, tax_percentage });
    } else {
      const {
        tax_id,
        tax_type,
        tax_percentage
      } = this.props.theRegions.taxes[0];
      this.setState({ tax_id, tax_type, tax_percentage });
    }
  };

  /**
   * ensures the user's card is charged whenever they place an order
   *
   * @param {Object} token
   *
   * @returns {void}
   */
  handlePayment = token => {
    const cartId = localStorage.getItem("cartId");
    const {
      theOrders: {
        totalAmountOfItemsInCart: { total_amount }
      }
    } = this.props;
    const { orderId } = this.state;
    const data = {
      stripeToken: token,
      oredr_id: orderId,
      description: "Some test description",
      amount: total_amount
    };
    const { createChargeOnCard } = this.props;
    createChargeOnCard(cartId, data);
  };

  handlePlaceOrder = event => {
    event.preventDefault();
    const cartId = localStorage.getItem("cartId");
    const { shipping_id, tax_id } = this.state;
    const { placeOrder } = this.props;
    const data = {
      cart_id: cartId,
      tax_id,
      shipping_id
    };
    if (this.props.theOrders.cartItems.length) {
      if (shipping_id === "1" || shipping_id === "0") {
        showErrorNotification("Please select a shipping region");
      } else {
        placeOrder(data);
      }
    } else {
      showErrorNotification("You don't have items in your cart");
    }
  };

  handleFetchOrderById = orderId => {
    const { getOrderById, getOrderDetailById } = this.props;
    getOrderById(orderId);
    getOrderDetailById(orderId);
    this.setState({ orderId });
  };

  /**
   * renders the validation errors for input fields
   *
   * @param {string} errorKey
   * @param {string} message
   *
   * @returns {void}
   */
  inputErrors = (errorKey, message) => {
    this.setState({
      ...(this.state.errors[errorKey] = message)
    });
  };

  /**
   * logs out a user from the app by doing the following
   * 1. clears the users token from localStorage
   * 2. clears the users name from loaclStorage
   * 3. redirects user to the landing page
   *
   * @returns {void}
   */
  handleLogout = () => {
    localStorage.removeItem("jwt-token");
    localStorage.removeItem("user");
    const { pathname } = history.location;
    if (pathname === "/") {
      history.push("/home");
    } else {
      history.push("/");
    }
  };

  formatDate = date => {
    const firstFormat = new Date(date);
    const theDate = new Date(firstFormat);
    const year = theDate.getFullYear();
    const month = theDate.getMonth() + 1;
    const day = theDate.getDate();
    const formatted = year + "-" + month + "-" + day;
    return formatted;
  };

  /**
   * logs out a user from the app by doing the following
   * 1. clears the users token from localStorage
   * 2. clears the users name from loaclStorage
   * 3. redirects user to the landing page
   *
   * @returns {void}
   */
  handleLogout = () => {
    localStorage.removeItem("jwt-token");
    localStorage.removeItem("user");
    const { pathname } = history.location;
    if (pathname === "/") {
      history.push("/home");
    } else {
      history.push("/");
    }
  };

  render() {
    const { isLoggedIn } = this.state;
    const {
      allDepartments,
      allDepartments: { departments },
      theRegions: { shippingRegions },
      theOrders: { myOrders, orderOfParticularId, orderDetails }
    } = this.props;
    const {
      cartItems,
      totalAmountOfItemsInCart: { total_amount }
    } = this.props.theOrders;
    const { email } = this.props.theRegions.customerDetails;
    return (
      <Fragment>
        <UpperNav
          allDepartments={allDepartments}
          isLoggedIn={isLoggedIn}
          totalAmount={total_amount}
          handleLogout={this.handleLogout}
          cartItems={cartItems}
        />
        <PlaceOrderComponent
          handleInputChange={this.handleInputChange}
          shippingRegions={shippingRegions}
          state={this.state}
          handlePlaceOrder={this.handlePlaceOrder}
          myOrders={myOrders}
          handleFetchOrderById={this.handleFetchOrderById}
          orderOfParticularId={orderOfParticularId}
          orderDetails={orderDetails}
          formatDate={this.formatDate}
          handlePayment={this.handlePayment}
          email={email}
          cartItems={cartItems}
        />
        <Footer departments={departments} />
      </Fragment>
    );
  }
}
const mapStateToProps = state => ({
  allDepartments: state.departments,
  theRegions: state.customers,
  theOrders: state.itemsInCart
});
export default connect(
  mapStateToProps,
  {
    getDepartments,
    getShippingRegions,
    getTaxes,
    placeOrder,
    getAllMyOrders,
    getOrderById,
    getOrderDetailById,
    getTotalAmount,
    getItemsInCart,
    createChargeOnCard,
    getCustomersDetails
  }
)(PlaceOrder);
