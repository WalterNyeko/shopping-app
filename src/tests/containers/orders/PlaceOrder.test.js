import React from "react";
import { shallow } from "enzyme";
import { PlaceOrder } from "../../../containers/orders/PlaceOrder";
import PlaceOrderComponent from "../../../components/orders/PlaceOrder";

describe("PlaceOrder", () => {
  const props = {
    getDepartments: jest.fn(),
    getShippingRegions: jest.fn(),
    getTaxes: jest.fn(),
    getAllMyOrders: jest.fn(),
    getTotalAmount: jest.fn(),
    getItemsInCart: jest.fn(),
    placeOrder: jest.fn(),
    getOrderById: jest.fn(),
    getOrderDetailById: jest.fn(),
    getCustomersDetails: jest.fn(),
    theOrders: {
      cartItems: [{}],
      totalAmountOfItemsInCart: { total_amount: 10 },
      myOrders: [{}],
      orderOfParticularId: [{}],
      orderDetails: {}
    },
    allDepartments: {
      departments: [{}]
    },
    theRegions: {
      shippingRegions: [{}],
      customerDetails: {
        email: "test@gmail.com"
      },
      taxes: [
        { tax_id: 1, tax_type: "Test", taxt_percentage: "2.5" },
        { tax_id: 2, tax_type: "Test 2", taxt_percentage: "3.5" }
      ]
    },
    createChargeOnCard: jest.fn(),
    cartItems: [{}]
  };
  const wrapper = shallow(<PlaceOrder {...props} />);
  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should call handleInputChange", () => {
    const spy = jest.spyOn(wrapper.instance(), "handleInputChange");
    const event = { target: { name: "shipping_id", value: "1" } };
    wrapper.instance().populateStateImmediately();
    wrapper.instance().handleInputChange(event);
    expect(spy).toHaveBeenCalled();
  });

  it("should call placeOrder() when handlePlaceOrder() is called", () => {
    const spy = jest.spyOn(wrapper.instance().props, "placeOrder");
    const event = {
      preventDefault: jest.fn()
    };
    wrapper.setState({ shipping_id: "2" });
    wrapper.instance().handlePlaceOrder(event);
    expect(spy).toHaveBeenCalled();
  });

  it("should not call placeOrder() when handlePlaceOrder() is called and shipping_id == 0", () => {
    const spy = jest.spyOn(wrapper.instance().props, "placeOrder");
    const event = {
      preventDefault: jest.fn()
    };
    wrapper.setState({ shipping_id: "0" });
    wrapper.instance().handlePlaceOrder(event);
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should not call placeOrder() if cart is empty", () => {
    wrapper.update();
    const spy = jest.spyOn(wrapper.instance().props, "placeOrder");
    const event = {
      preventDefault: jest.fn()
    };
    wrapper.setProps({
      theOrders: {
        cartItems: [],
        totalAmountOfItemsInCart: { total_amount: 2 }
      }
    });
    wrapper.setState({ shipping_id: "2" });
    wrapper.instance().handlePlaceOrder(event);
    wrapper.instance().handleFetchOrderById(1);
    wrapper.instance().inputErrors("test", "test complain");
    wrapper.instance().handleLogout();
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should format date correctly", () => {
    const formattedDate = wrapper
      .instance()
      .formatDate("2019-06-17T01:12:13.000Z");
    expect(formattedDate).toEqual("2019-6-17");
  });

  it("should render the component well", () => {
    const componentProps = {
      state: {
        errors: {}
      },
      orderDetails: {
        total_amount: 10
      },
      shippingRegions: [{}],
      handleInputChange: jest.fn(),
      handlePlaceOrder: jest.fn(),
      myOrders: [{}],
      handleFetchOrderById: jest.fn(),
      orderOfParticularId: [{}],
      orderDetails: jest.fn(),
      formatDate: jest.fn(),
      cartItems: [{}]
    };
    const componentWrapper = shallow(
      <PlaceOrderComponent {...componentProps} />
    );
    expect(componentWrapper).toMatchSnapshot();
  });

  it("should call handlePayment() when the function is called", () => {
    const spy = jest.spyOn(wrapper.instance(), "handlePayment");
    wrapper.instance().handlePayment();
    expect(spy).toHaveBeenCalled();
  });
});
