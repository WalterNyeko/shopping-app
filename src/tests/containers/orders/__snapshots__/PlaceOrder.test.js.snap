// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceOrder should render correctly 1`] = `ShallowWrapper {}`;

exports[`PlaceOrder should render the component well 1`] = `ShallowWrapper {}`;
